dattri/algorithm/data_shapley.py

```diff
@@ -0,0 +1,135 @@
+                    coord1 = train_batch[0]
+                    coord2 = test_batch[0]
+                    return torch.cdist(coord1, coord2)
+                ```.
+        """
+        self.k = k
+        self.distance_func = distance_func
+
+    def cache(self) -> None:
+        """Precompute and cache some values for efficiency."""
+
+    def attribute(
+        self,
+        train_dataloader: torch.utils.data.DataLoader,
+        test_dataloader: torch.utils.data.DataLoader,
+        train_labels: List[int],
+        test_labels: List[int],
+    ) -> None:
+        """Calculate the shapley values of the training set on each test sample.
+
+        Args:
+            train_dataloader (torch.utils.data.DataLoader): The dataloader for
+                training samples to calculate the shapley values. The dataloader
+                should not be shuffled.
+            test_dataloader (torch.utils.data.DataLoader): The dataloader for
+                test samples to calculate the shapley values. The dataloader
+                should not be shuffled.
+            train_labels: (List[int]): The list of training labels, with the same
+                size and order of the training dataset.
+            test_labels: (List[int]): The list of test labels, with the same
+                size and order of the test dataset.
+
+        Returns:
+            Tensor: The shapley values of the training set on the test set, with
+                the shape of (num_train_samples, num_test_samples).
+        """
+        _check_shuffle(test_dataloader)
+        _check_shuffle(train_dataloader)
+
+        num_train_samples = len(train_dataloader.sampler)
+        num_test_samples = len(test_dataloader.sampler)
+
+        dist_matrix = torch.zeros(
+            size=(num_test_samples, num_train_samples),
+        )
+
+        shapley_values = torch.zeros(
+            size=(num_test_samples, num_train_samples),
+        )
+
+        for test_batch_idx, test_batch_data in enumerate(
+            tqdm(
+                test_dataloader,
+                desc="fitting KNN...",
+                leave=False,
+            ),
+        ):
+            for train_batch_idx, train_batch_data in enumerate(
+                train_dataloader,
+            ):
+                partial_dist = self.distance_func(test_batch_data, train_batch_data)
+
+                # results position based on batch info
+                col_st = train_batch_idx * train_dataloader.batch_size
+                col_ed = min(
+                    (train_batch_idx + 1) * train_dataloader.batch_size,
+                    len(train_dataloader.sampler),
+                )
+
+                row_st = test_batch_idx * test_dataloader.batch_size
+                row_ed = min(
+                    (test_batch_idx + 1) * test_dataloader.batch_size,
+                    len(test_dataloader.sampler),
+                )
+
+                dist_matrix[row_st:row_ed, col_st:col_ed] += partial_dist
```

Comment:
do we need to take batch size into consideration when computing this distance matrix?
it should be faster to directly call `self.distance_func` on the two lists?
(although this part should be quite fast even computed in this batch-style)

Reply:
Actually I am concerned about the memory issue of calculating the distance matrix, because the data can be something other than tensor, and the user may also include some preprocessing in their distance computation. I don't know whether concatenating everything into a list would cause oom issues.

Reply:
The batch version sounds good to me.

Reply:
sure, then it should better to just keep the batch version.